### ISSUE TITLE
Cache update may fall behind flushes leading to memory overflow

### ISSUE THREAD

**Issue by tgrabiec** [scylladb/scylla#1942]
Until partitions are moved to cache, the memory is not really evictable yet. If cache update lags behind flushes, non-evictable memory may accumulate and lead to bad_allocs.

This is clearly visible when cache update concurrency is limited to 1 on a saturated CPU.  (graph below: blue - total memory, yellow - virtual dirty, green - real dirty):

![dirty overflow](https://cloud.githubusercontent.com/assets/283695/21257175/e7cdcf8a-c376-11e6-98d2-cc1bb3050761.png)

**Comment by tgrabiec** [comment]
Refs #1943

**Comment by tgrabiec** [comment]
This diagram depicts how things work currently:

![dirty memory](https://user-images.githubusercontent.com/283695/30901595-bb96140e-a368-11e7-97ae-225590801e60.png)

The sum of "Flushed" and "Not flushes" regions is non-evictable memory, which we want to limit to 1/2 of the shard's memory. However, currently only "Not flushed" (aka virtual dirty) is limited, by stopping writes when it reaches 1/4 of shard's memory (aka virtual dirty hard threshold). The "Flushed" part can run away if cache update is slow.

Another problem is that "real dirty" doesn't account for memory in the memtable which has started to be moved to cache. This is the first thing to fix. After that, "real dirty" will include "Flushed"+"Not flushed", which we want to limit.

Then we should make sure that writes stop not only when we hit the virtual dirty limit ("Not flushed") but also when we hit the real dirty limit ("Flushed"+"Not flushed"). In case incoming writes are overly fast, they will eventually move at the speed of either the flusher, or cache updater, whichever is slower.

**Comment by avikivity** [comment]
Doesn't that mean we need to move to triple-buffering (with memtable size limited to 1/6 of shard memory)?

**Comment by avikivity** [comment]
Let's move to single-buffering instead. As soon as we know a partition (or better, row) is on disk, update in-memory structures allowing as to find it (= summary if the index for that partition is on disk, some other structure if not), move the partition/row to cache. We don't need to fsync since the data is in commitlog.

Suppose the memtable flush failed. We can continue reading from the partially written sstable while writing the new sstable. So the flush process becomes (merge data from (partially written sstable if any, memtable) -> new sstable).

On crash, we read the commitlog anyway.

We'd need to update the sstable's size in real time during writing, add the temporary index structure, and open the files in read/write mode. But it seems like a reasonably small change.

/cc @glommer @raphaelsc

**Comment by avikivity** [comment]
Besides solving the bug, this would double the size of tier 0 sstables, improving performance.

**Comment by glommer** [comment]
I have proposed a similar solution a long time ago. What ended up convincing us not to do it was your very argument that memory can't really be freed when moved to the cache. As a consequence, it is hard to initiate a second memtable flush: the sum of two memtables in memory now would now take over the entire memory in the worst case.

Am I missing something ?

**Comment by avikivity** [comment]
I outlined above how it can really be freed.

**Comment by glommer** [comment]
I still don't understand, since that was exactly what I proposed back then. I need to go see if I can find the logs for that discussion, but if I recall correctly the problem seemed to be that freeing a partition doesn't necessarily free the segment. In the worst case we can end up with many partitions freed but not segment really freed.

**Comment by avikivity** [comment]
You're right, moving the partition/row to cache still leaves it in an unevictable region. The only way to really reduce the size is to copy the data. Perhaps that's what we should do. It's expensive, but it's cheaper than an extra compaction (doubling the tier 0 sstable size saves us just half a compaction, but perhaps more if we get more overwrites).

**Comment by avikivity** [comment]
Not really sure about unevictability any more. It's partially unevictable but that's what it's all about! And it is compactible.

**Comment by gleb-cloudius** [comment]
Why can't we just drop fully flushed sstable even if it is not yet fully moved into the cache?

**Comment by avikivity** [comment]
We can.

**Comment by slivne** [comment]
> "Why can't we just drop fully flushed sstable even if it is not yet fully moved into the cache?"

Trying to understand what "drop" means here - is the memtable still around - we have to continue reading from it till its fully merged into the cache.

The cache can hold "stale data" - that does not include the flushed sstable and we have to read from the memtable (or invalidate the related info in the cache - that we currently don't do).

**Comment by gleb-cloudius** [comment]
On Sun, Oct 01, 2017 at 12:51:09AM -0700, Shlomi Livne wrote:
> > "Why can't we just drop fully flushed sstable even if it is not yet fully moved into the cache?"
> 
> Trying to understand what "drop" means here - is the memtable still around - we have to continue reading from it till its fully merged into the cache.
> 
Drop means free its memory.

> The cache can hold "stale data" - that does not include the flushed sstable and we have to read from the memtable (or invalidate the related info in the cache - that we currently don't do).
> 
Good point, but we can still do it if the cache is empty. So if we are
low on memory we evict the cache and if memory goal is still not reached
we can free already flushed memtables instead of moving them to the
cache and evicting from there.

--
			Gleb.

**Comment by tgrabiec** [comment]
2017-09-30 18:13 GMT+02:00 Avi Kivity <notifications@github.com>:

> Doesn't that mean we need to move to triple-buffering (with memtable size
> limited to 1/6 of shard memory)?
>

I don't think so, why? With regards to buffering, all stays the same. This
is strictly an improvement of the current situation. If cache update
doesn't fall behind, everything behaves exactly as currently.

**Comment by tgrabiec** [comment]
2017-09-30 18:20 GMT+02:00 Avi Kivity <notifications@github.com>:

> Let's move to single-buffering instead. As soon as we know a partition (or
> better, row) is on disk, update in-memory structures allowing as to find it
> (= summary if the index for that partition is on disk, some other structure
> if not), move the partition/row to cache. We don't need to fsync since the
> data is in commitlog.
>
> Suppose the memtable flush failed. We can continue reading from the
> partially written sstable while writing the new sstable. So the flush
> process becomes (merge data from (partially written sstable if any,
> memtable) -> new sstable).
>
> On crash, we read the commitlog anyway.
>
> We'd need to update the sstable's size in real time during writing, add
> the temporary index structure, and open the files in read/write mode. But
> it seems like a reasonably small change.
>

This is an improvement, but I think orthogonal to this problem. Even if we
flush in small increments, we need to move those increments to cache, and
need to throttle writes if cache update is lagging behind.

**Comment by avikivity** [comment]
> I don't think so, why? With regards to buffering, all stays the same. This
is strictly an improvement of the current situation. If cache update
doesn't fall behind, everything behaves exactly as currently.

If we have three active memtables controlled by a controller, in the worst case all three are full. If their size is mem/4, then they can occupy 3/4 of all memory.

**Comment by avikivity** [comment]
> This is an improvement, but I think orthogonal to this problem. Even if we
flush in small increments, we need to move those increments to cache, and
need to throttle writes if cache update is lagging behind.

We already update virtual dirty based on those increments. We just change the definition of "flushed" from "written to sstable" to "written to sstable and content moved to cache".

**Comment by avikivity** [comment]
It can even be performed by the same thread, rather than having an additional background thread. We can have the sstable writer produce a stream of mutations (it acts as a pass-through reader) that have been flushed.

**Comment by tgrabiec** [comment]
2017-10-02 10:44 GMT+02:00 Avi Kivity <notifications@github.com>:

> I don't think so, why? With regards to buffering, all stays the same. This
> is strictly an improvement of the current situation. If cache update
> doesn't fall behind, everything behaves exactly as currently.
>
> If we have three active memtables controlled by a controller, in the worst
> case all three are full. If their size is mem/4, then they can occupy 3/4
> of all memory.
>

The amount of unevictable memory (purple+yellow on the diagram) is always
at most 1/2 of shard's memory, because the controller stops admitting
writes when it's reached. Note that unevictable size of a memtable
decreases as cache update progresses over it.

**Comment by avikivity** [comment]
I see, thanks.

**Comment by tgrabiec** [comment]
2017-10-02 10:45 GMT+02:00 Avi Kivity <notifications@github.com>:

> This is an improvement, but I think orthogonal to this problem. Even if we
> flush in small increments, we need to move those increments to cache, and
> need to throttle writes if cache update is lagging behind.
>
> We already update virtual dirty based on those increments. We just change
> the definition of "flushed" from "written to sstable" to "written to
> sstable and content moved to cache".
>

We could do that, but it may be still better to update cache in parallel so
that it's latency does not add to write-perceived latency. Cache update can
be hidden behind I/O to disk.

**Comment by avikivity** [comment]
It's cpu work, so no real parallelism is possible. Disk work is already parallel via write-behind.

**Comment by glommer** [comment]
Folks,

Why can't we just make the current dirty group a leaf group, and add a parent that controls for real dirty ? We would set the threshold for the parent to the real dirty limit, and if it were to block, the child would block as well.

We can then keep track of memory that is supposedly freed but blocked by cache updates and use that as input for an update_cache controller.

**Comment by avikivity** [comment]
I guess, if the cache-update memtable size + the active memtable size are limited to below half the memtable space, then we have strict double-buffering.

**Comment by avikivity** [comment]
So the potential from the optimization is 2X run size, and just a 0.5 compaction (assuming min_threshold = 4).
